proposals/p1928.md

```diff
@@ -0,0 +1,163 @@
+# Arrays
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1928)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+Arrays are base component of almost of programming languages. In the current
+stage of Carbon development, arrays are missed.
+
+## Background
+
+An array is a set of scalar data placed in contiguous memory locations, all of
+the same type and type parameters. The scalar data that make up an array are the
+array elements which can be referenced by unique indexes.
+
+A nested array is an array of arrays which may be non-continuous in memory.
+
+A multidimensional array is an array with more than two dimensions which is
+continuous in memory.
+
+## Proposal
+
+This proposal introduces array internal representation, its declaration, and
+indexing.
+
+## Details
+
+### Internal representation
+
+For simplifying usage and raising safety of array usage, internal representation
+of arrays contains additional metadata:
+```cpp
+  struct {
+    int32_t dimensions;
+    int32_t * dimension_size;
+    void * data;
```

Comment:
What's the ownership model for this data?

Reply:
This data will be destroyed at the end of scope (or when calling destructors of base scope, like classes). Later, via standard library, it would be possible to destroy object earlier.

Reply:
Is this really a pointer to the data or is the data stored in-line after the header information, eg `T * n`, where `n` is the product of all the `dimension_size`s?

Reply:
It is a pointer to the data. For example, in C++, to allocate some memory for array, then in Carbon code, to build header over array.

Reply:
I suspect the most common uses for arrays in C++ today are to avoid the allocation intrinsic to `std::vector`. I think Carbon will want to have a guaranteed non-allocating array type.

Reply:
> I think Carbon will want to have a guaranteed non-allocating array type.

No doubt, but I interpret this proposal as saying "the non-allocating array type gets some other type syntax" rather than "the non-allocating array type doesn't exist", although the proposal should probably be explicit about that.

That said, I think the non-allocating type should get this syntax, because I think we only have syntactic "room" for one built-in array type syntax, and all the other array types will need to be spelled like library types (e.g. `StaticArray(T, N)`). That being the case, I think it would be surprising and confusing if the array type that "looks built in" were not the lowest-level type.

(I think there will actually be two lowest-level array types -- the other one is what I referred to [here](https://github.com/carbon-language/carbon-lang/discussions/1061) as `RuntimeSizedArray`. That one is much weirder and more awkward to use than either `std::array` or `std::vector`, so it definitely doesn't get to be the one with a built-in syntax.)

Reply:
> we only have syntactic "room" for one built-in array type syntax
I don't think this has to be true at all, but I don't mind at all if we go that way.

Reply:
> > we only have syntactic "room" for one built-in array type syntax
> 
> I don't think this has to be true at all, but I don't mind at all if we go that way.

Well, I'll put it this way: I think we should presume this is the only built-in array syntax, unless this PR also proposes a syntax for statically-sized arrays, so we assess how the language looks with both of them.

Reply:
> unless this PR also proposes a syntax for statically-sized arrays

I hope syntax for statically-sized and runtime-sized arrays will be close. Probably, the simplest change is to replace numbers with colon.